### PR TITLE
Feature: Reworked presence notifications.

### PR DESCRIPTION
**Pull request by ppenchev-storpool**
Rework the way the StorPool charms (cinder-storpool and storpool-block) notify their own peers and the other charm about the readiness of the units on specific machines and with specific StorPool IDs.

### CODE REVIEW

**Comment by ryan-beisner** [comment]
Unit tests are failing on this branch:

```
pep8 installed: charmhelpers==0.18.5,charms.reactive==0.6.2,cliff==2.11.0,cmd2==0.8.1,extras==1.0.0,fixtures==3.0.0,flake8==3.5.0,future==0.16.0,Jinja2==2.10,linecache2==1.0.0,MarkupSafe==1.0,mccabe==0.6.1,mock==2.0.0,netaddr==0.7.19,os-testr==1.0.0,pbr==3.1.1,pkg-resources==0.0.0,prettytable==0.7.2,pyaml==17.12.1,pycodestyle==2.3.1,pyflakes==1.6.0,pyparsing==2.2.0,pyperclip==1.6.0,python-mimeparse==1.6.0,python-subunit==1.2.0,PyYAML==3.12,six==1.11.0,stestr==2.0.0,stevedore==1.28.0,Tempita==0.5.2,testtools==2.3.0,traceback2==1.4.0,unittest2==1.1.0,voluptuous==0.11.1
pep8 runtests: PYTHONHASHSEED='771664293'
pep8 runtests: commands[0] | flake8 reactive
pep8 runtests: commands[1] | flake8 --ignore=E402 unit_tests
unit_tests_35 create: /home/ubuntu/git/storpool-charms/storpool-charms/charms/charm-storpool-block/.tox/unit_tests_35
unit_tests_35 installdeps: -r/home/ubuntu/git/storpool-charms/storpool-charms/charms/charm-storpool-block/test-requirements.txt
unit_tests_35 installed: charmhelpers==0.18.5,charms.reactive==0.6.2,cliff==2.11.0,cmd2==0.8.1,extras==1.0.0,fixtures==3.0.0,flake8==3.5.0,future==0.16.0,Jinja2==2.10,linecache2==1.0.0,MarkupSafe==1.0,mccabe==0.6.1,mock==2.0.0,netaddr==0.7.19,os-testr==1.0.0,pbr==3.1.1,pkg-resources==0.0.0,prettytable==0.7.2,pyaml==17.12.1,pycodestyle==2.3.1,pyflakes==1.6.0,pyparsing==2.2.0,pyperclip==1.6.0,python-mimeparse==1.6.0,python-subunit==1.2.0,PyYAML==3.12,six==1.11.0,stestr==2.0.0,stevedore==1.28.0,Tempita==0.5.2,testtools==2.3.0,traceback2==1.4.0,unittest2==1.1.0,voluptuous==0.11.1
unit_tests_35 runtests: PYTHONHASHSEED='771664293'
unit_tests_35 runtests: commands[0] | ostestr
{0} unittest2.loader._FailedTest.unit_tests.test_block_charm [0.000477s] ... FAILED

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    b'Traceback (most recent call last):'
    b'ImportError: Failed to import test module: unit_tests.test_block_charm'
    b'Traceback (most recent call last):'
    b'  File "/home/ubuntu/git/storpool-charms/storpool-charms/charms/charm-storpool-block/.tox/unit_tests_35/lib/python3.5/site-packages/unittest2/loader.py", line 456, in _find_test_path'
    b'    module = self._get_module_from_name(name)'
    b'  File "/home/ubuntu/git/storpool-charms/storpool-charms/charms/charm-storpool-block/.tox/unit_tests_35/lib/python3.5/site-packages/unittest2/loader.py", line 395, in _get_module_from_name'
    b'    __import__(name)'
    b'  File "/home/ubuntu/git/storpool-charms/storpool-charms/charms/charm-storpool-block/unit_tests/test_block_charm.py", line 117, in <module>'
    b'    from reactive import storpool_block_charm as testee'
    b'  File "/home/ubuntu/git/storpool-charms/storpool-charms/charms/charm-storpool-block/reactive/storpool_block_charm.py", line 30, in <module>'
    b'    from spcharms import kvdata'
    b"ImportError: cannot import name 'kvdata'"
    b''
    b''

==============================
Failed 1 tests - output below:
==============================

unittest2.loader._FailedTest.unit_tests.test_block_charm
--------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    b'Traceback (most recent call last):'
    b'ImportError: Failed to import test module: unit_tests.test_block_charm'
    b'Traceback (most recent call last):'
    b'  File "/home/ubuntu/git/storpool-charms/storpool-charms/charms/charm-storpool-block/.tox/unit_tests_35/lib/python3.5/site-packages/unittest2/loader.py", line 456, in _find_test_path'
    b'    module = self._get_module_from_name(name)'
    b'  File "/home/ubuntu/git/storpool-charms/storpool-charms/charms/charm-storpool-block/.tox/unit_tests_35/lib/python3.5/site-packages/unittest2/loader.py", line 395, in _get_module_from_name'
    b'    __import__(name)'
    b'  File "/home/ubuntu/git/storpool-charms/storpool-charms/charms/charm-storpool-block/unit_tests/test_block_charm.py", line 117, in <module>'
    b'    from reactive import storpool_block_charm as testee'
    b'  File "/home/ubuntu/git/storpool-charms/storpool-charms/charms/charm-storpool-block/reactive/storpool_block_charm.py", line 30, in <module>'
    b'    from spcharms import kvdata'
    b"ImportError: cannot import name 'kvdata'"
    b''
    b''


======
Totals
======
Ran: 1 tests in 0.4912 sec.
 - Passed: 0
 - Skipped: 0
 - Expected Fail: 0
 - Unexpected Success: 0
 - Failed: 1
Sum of execute time for each test: 0.0005 sec.

==============
Worker Balance
==============
 - Worker 0 (1 tests) => 0:00:00.000477

Slowest Tests:
Test id                                                   Runtime (s)
--------------------------------------------------------  -----------
unittest2.loader._FailedTest.unit_tests.test_block_charm  0.000
ERROR: InvocationError: '/home/ubuntu/git/storpool-charms/storpool-charms/charms/charm-storpool-block/.tox/unit_tests_35/bin/ostestr'
___________________________________ summary ____________________________________
  pep8: commands succeeded
ERROR:   unit_tests_35: commands failed
```

**Comment by ajkavanagh** [comment]
Unit tests are now passing.

File: reactive/storpool_block_charm.py [hunk 1]
```diff
@@ -59,6 +59,8 @@ def install_setup():
     """
     spconfig.set_meta_config(None)
     spstatus.set_status_reset_handler('storpool-repo-add')
+    if hookenv.config().get('storpool_repo_url') is None:
```
Comment (ajkavanagh) [inline reactive/storpool_block_charm.py]: Note that this line can be written as:

```python
if hookenv.config('storpool_repo_url') is None:
    ...
```

i.e. the config() function supports asking for just a key from the config data.

**Comment by ajkavanagh** [comment]
It must be really hard to reason about this as there are so many states; is there a state transition diagram (or similar) that shows what states there are and how they are invoked.  I realise it's a dynamic flow, but it's really hard to follow what's going on here.
